# Project Report Notebook - Football Transfer Analytics

[Executive_SummaryAbstract](#Executive_Summary/Abstract)
<br>[Table of Contents Introduction](#Table_of_Contents_Introduction)
<br>[Analysis](#analysis)
<br>[Conclusion](#conclusion)
<br>[Appendix: Project Structure](#Project_Structure)

## Executive Summary/Abstract <a id='Executive_Summary/Abstract'></a>

<br>The question: What recipy a football player should follow in order to increase his market value the most?

<br>Purpose: We want to analyze what are the most important factors that might increase the value of a player when he transfers from one club to another. Hence, we built a network by taking 10 years of historical data and computed different centrality measurements in order to find out valuable information using machine learning algorithms.

- Network: Transfers of professional and semi-professional football players.
- Type: Directed and weighted network
- Source: [TransferMarket](https://www.transfermarkt.co.uk/).
- Modification: We have scrapped the clubs that have a total market value above 200 million euros. From these clubs, we downloaded the html files of their football players. This include the nationality, birthdate, transfer history, transfer fee etc.

## Table of Contents Introduction

In [7]:
from urllib.request import *
from bs4 import BeautifulSoup
import re
import pandas as pd
import numpy as np
import scipy.stats as stats
import pylab as pl
import requests
from geopy.geocoders import Nominatim
import json
import folium
import os
import time
from tempfile import TemporaryFile
from datetime import datetime
import matplotlib.pyplot as plt
from collections import Counter
from scipy import sparse, stats, spatial
import scipy.sparse.linalg
import networkx as nx
from pylab import rcParams
%matplotlib inline

# ML related libraries
from sklearn.linear_model import Lasso
import copy
from sklearn.externals import joblib
from sklearn.metrics import mean_absolute_error
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neural_network import MLPRegressor
import sklearn.svm as svm
from sklearn.ensemble import RandomForestRegressor
from sklearn.decomposition import PCA

from models import *
from helpers import *

import warnings
warnings.filterwarnings('ignore')

### Load data

In [2]:
ifile = open('transfers.csv')
df = pd.read_csv(ifile)

G = nx.read_graphml("transfers.graphml")

In [3]:
print(nx.info(G))

Name: 
Type: DiGraph
Number of nodes: 401
Number of edges: 9621
Average in degree:  23.9925
Average out degree:  23.9925


We can see that some of the 418 clubs were not present in the graph. That's because the players of these clubs did not transfer to other clubs given the time constraint of 5 years (or no one from the other clubs transfered to them).

In [4]:
df

from_club_id  to_club_id player_nationality  value_increase  \
0              1184         631            Belgium         8950000   
1               631          13            Belgium        -7750000   
2                13         631            Belgium        23800000   
3              1084         281          Argentina         7100000   
4               281         631          Argentina        -8000000   
5              1075        1085           Portugal               0   
6              1085        1075           Portugal         1500000   
7              1075         252           Portugal         3000000   
8               252         294           Portugal         2000000   
9               294         252           Portugal        -2500000   
10              252        6890           Portugal               0   
11             6890         252           Portugal         -300000   
12              252        1075           Portugal               0   
13             1075         252           Portugal         -200000   
14              631         499            Croatia        -1500000   
15              499         631            Croatia         -500000   
16              631        2420            Croatia               0   
17             2420         631            Croatia               0   
18              631       29228            Croatia               0   
19            29228         631            Croatia         -100000   
20             2125         294             Brazil               0   
21              294        2125             Brazil         1500000   
22             2125         294             Brazil        -1000000   
23              294         631             Brazil        24500000   
24              631         583             Brazil        24500000   
25              583         631             Brazil       -14500000   
26               79          12            Germany         4000000   
27               12          79            Germany         8000000   
28               79          12            Germany        -3000000   
29               12         631            Germany        26000000   
...             ...         ...                ...             ...   
18366          1062        8344              Japan         -100000   
18367         16032        8344              Japan               0   
18368         22177        8344              Japan          450000   
18369           776        1134             Brazil               0   
18370          1134         776             Brazil               0   
18371           776        8344             Brazil               0   
18372           609        2646             Brazil               0   
18373          2646         609             Brazil          650000   
18374           609        8344             Brazil          480000   
18375          7452       22177              Japan               0   
18376         22177        8344              Japan          500000   
18377           596       10999              Japan               0   
18378         10999         596              Japan          250000   
18379           596       10999              Japan         -250000   
18380          5677         221             Brazil          110000   
18381           221       21829             Brazil         -110000   
18382         21829        3187             Brazil               0   
18383          2863       10999             Brazil               0   
18384          3828       10999              Japan               0   
18385           596       10999              Japan         -350000   
18386          7178         210             Brazil          300000   
18387           210        7178             Brazil               0   
18388          7178        7209             Brazil               0   
18389          7209        7178             Brazil               0   
18390          7178         596             Brazil         -300000   
18391          40

### Predicting player value increase

In [5]:
SEED = 20

def build_models(list_of_network_features, normalize = False):
    for columns_to_keep in list_of_network_features:
        x_train, y_train = get_data(columns_to_keep)
        if normalize:
            x_train = normalize_feat(x_train)
        
        print ("Build models that include {} network measures".format(columns_to_keep))
        
        # Build Lasso regression
        lasso_regressor = Lasso_Regression(alpha=1e4, seed=SEED)
        # Get cross validation error
        mean_error_lasso, std_error_lasso = cross_validation(model_ori=lasso_regressor, input=x_train, labels=y_train, K=5)
        print('* Lasso cross validation error mean: \t\t{}, \t\tstd: {}'.format(int(mean_error_lasso), int(std_error_lasso)))

        # Build KNN regressor
        knn_regressor = KNN()
        # Get cross validation error
        mean_error_knn, std_error_knn = cross_validation(model_ori=knn_regressor, input=x_train, labels=y_train, K=5)
        print('* KNN cross validation error mean: \t\t{}, \t\tstd: {}'.format(int(mean_error_knn), int(std_error_knn)))

        # Build MLP regressor
        mlp_regressor = MLP(seed=SEED, solver='adam', alpha=1e-5, hidden_layers=(25, 25), lr=1e-4, max_iter=100000)
        # Get cross validation error
        mean_error_mlp, std_error_mlp = cross_validation(model_ori=mlp_regressor, input=x_train, labels=y_train, K=5)
        print('* MLP cross validation error mean: \t\t{}, \t\tstd: {}'.format(int(mean_error_mlp), int(std_error_mlp)))
        
        # Build SVR regressor
        svr_regressor = SVR(kernel='linear', seed=SEED)
        # Get cross validation error
        mean_error_svm, std_error_svm = cross_validation(model_ori=svr_regressor, input=x_train, labels=y_train, K=5)
        print('* SVR cross validation error mean: \t\t{}, \t\tstd: {}'.format(int(mean_error_svm), int(std_error_svm)))

        # Build Random Forest Regressor
        random_forest_regressor = Random_Forest(n_estimators=100, max_depth=20,seed=SEED)
        # Get cross validation error
        mean_error_rf, std_error_rf = cross_validation(model_ori=random_forest_regressor,input=x_train, labels=y_train, K=5)
        print('* Random Forest cross validation error mean: \t{}, \t\tstd: {} \n'.format(int(mean_error_rf), int(std_error_rf)))

### Feature evaluation

In [8]:
build_models([['to_closeness','to_in_degree'],[]])

Build models that include ['to_closeness', 'to_in_degree'] network measures
* Lasso cross validation error mean: 		1795623, 		std: 994607
* KNN cross validation error mean: 		1680808, 		std: 1067207
* MLP cross validation error mean: 		1743559, 		std: 1037453
* SVR cross validation error mean: 		2397499, 		std: 1390480
* Random Forest cross validation error mean: 	1563739, 		std: 1068441 

Build models that include [] network measures
* Lasso cross validation error mean: 		1795680, 		std: 994646
* KNN cross validation error mean: 		1680814, 		std: 1067209
* MLP cross validation error mean: 		1766269, 		std: 1017408
* SVR cross validation error mean: 		2397499, 		std: 1390480
* Random Forest cross validation error mean: 	1571424, 		std: 1062634 



Error for 'to_closeness' and 'from_in_degree': * Random Forest cross validation error mean: 	1559847, 		std: 1054189 
<br>Error for no centrality: * Random Forest cross validation error mean: 	1571424, 		std: 1062634
<br>
<br>Error for 'to_closeness' and 'to_in_degree': * Random Forest cross validation error mean: 	1563739, 		std: 1068441 
<br>Error for no centrality: * Random Forest cross validation error mean: 	1571424, 		std: 1062634 

## Analysis

## Conclusion

## Appendix: Project Structure